clang/docs/analyzer/user-docs/FAQ.rst

```diff
@@ -0,0 +1,208 @@
+
+If your class only uses an instance variable for part of its lifetime, it may maintain an invariant guaranteeing that the instance variable is always released before -dealloc. In this case, you can silence a warning about a missing release by either adding ``assert(_ivar == nil)`` or an explicit release ``[_ivar release]`` (which will be a no-op when the variable is nil) in -dealloc.
+
+Deciding Nullability
+--------------------
+
+Q: How do I decide whether a method's return type should be _Nullable or _Nonnull?
+
+Depending on the implementation of the method, this puts you in one of five situations:
+
+1. You actually never return nil.
+2. You do return nil sometimes, and callers are supposed to handle that. This includes cases where your method is documented to return nil given certain inputs.
+3. You return nil based on some external condition (such as an out-of-memory error), but the client can't do anything about it either.
+4. You return nil only when the caller passes input documented to be invalid. That means it's the client's fault.
+5. You return nil in some totally undocumented case.
+
+In (1) you should annotate the method as returning a ``_Nonnull`` object.
+
+In (2) the method should be marked ``_Nullable``.
+
+In (3) you should probably annotate the method ``_Nonnull``. Why? Because no callers will actually check for nil, given that they can't do anything about the situation and don't know what went wrong. At this point things have gone so poorly that there's basically no way to recover.
+
+The least happy case is (4) because the resulting program will almost certainly either crash or just silently do the wrong thing. If this is a new method or you control the callers, you can use ``NSParameterAssert()`` (or the equivalent) to check the precondition and remove the nil return. But if you don't control the callers and they rely on this behavior, you should return mark the method ``_Nonnull`` and return nil cast to _Nonnull anyway.
+
+If you're in (5), document it, then figure out if you're now in (2), (3), or (4).
+
+Intentional Nullability Violation
+---------------------------------
+
+Q: How do I tell the analyzer that I am intentionally violating nullability?
+
+In some cases, it may make sense for methods to intentionally violate nullability. For example, your method may — for reasons of backward compatibility — chose to return nil and log an error message in a method with a non-null return type when the client violated a documented precondition rather than check the precondition with ``NSAssert()``. In these cases, you can suppress the analyzer warning with a cast:
+
+.. code-block:: objc
+
+   return (id _Nonnull)nil;
+
+Note that this cast does not affect code generation.
+
+Ensuring Loop Body Execution
+----------------------------
+
+Q: The analyzer assumes that a loop body is never entered. How can I tell it that the loop body will be entered at least once?
```

Comment:
I'd say that this is a bug (unwanted counter-intuitive behavior) and I hope that a generalized version of https://github.com/llvm/llvm-project/pull/109804 will eventually eliminate this question.